pkg/operators/ebpf/tracer.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpfoperator
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/btf"
+	"github.com/cilium/ebpf/perf"
+	"github.com/cilium/ebpf/ringbuf"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	metadatav1 "github.com/inspektor-gadget/inspektor-gadget/pkg/metadata/v1"
+)
+
+type Tracer struct {
+	metadatav1.Tracer
+
+	ds       datasource.DataSource
+	accessor datasource.FieldAccessor
+
+	mapType       ebpf.MapType
+	ringbufReader *ringbuf.Reader
+	perfReader    *perf.Reader
+}
+
+func validateTraceMap(traceMap *ebpf.MapSpec) error {
+	if traceMap.Type != ebpf.RingBuf && traceMap.Type != ebpf.PerfEventArray {
+		return fmt.Errorf("map %q has a wrong type, expected: ringbuf or perf event array, got: %s",
+			traceMap.Name, traceMap.Type.String())
+	}
+	return nil
+}
+
+func (i *ebpfInstance) populateTracer(t btf.Type, varName string) error {
+	i.logger.Debugf("populating tracer %q", varName)
+
+	parts := strings.Split(varName, typeSplitter)
+	if len(parts) != 3 {
+		return fmt.Errorf("invalid tracer info: %q", varName)
+	}
+
+	name := parts[0]
+	mapName := parts[1]
+	mapType := parts[2]
+
+	i.logger.Debugf("> name     : %q", name)
+	i.logger.Debugf("> map name : %q", mapName)
+	i.logger.Debugf("> type name: %q", mapType)
+
+	tracerConfig := i.config.Sub("tracers." + name)
+	if tracerConfig != nil {
+		if configMapName := tracerConfig.GetString("mapName"); configMapName != "" && configMapName != mapName {
+			return fmt.Errorf("mapName in tracer metadata for %q does not match metadata", varName)
+		}
+		if configMapType := tracerConfig.GetString("mapType"); configMapType != "" && configMapType != mapType {
+			return fmt.Errorf("mapName in tracer metadata for %q does not match metadata", varName)
+		}
+		i.logger.Debugf("> successfully validated with metadata")
+	}
+
+	if _, ok := i.tracers[name]; ok {
+		i.logger.Debugf("tracer %q already defined, skipping", name)
+		return nil
+	}
```

Comment:
I think in the original code this was done to prevent populating something that was already defined by metadata (and instead just doing some validation like in the lines above). But right now that case isn't handled here (discussion in #2560), so it shouldn't happen.